pulsar-broker/src/main/java/org/apache/pulsar/broker/service/TopicEventsDispatcher.java

```diff
@@ -0,0 +1,136 @@
+        Arrays.stream(listeners)
+                .filter(x -> x != null)
+                .forEach(topicEventListeners::remove);
+    }
+
+    /**
+     * Dispatches notification to all currently added listeners.
+     * @param topic
+     * @param event
+     * @param stage
+     */
+    public void notify(String topic,
+                       TopicEventsListener.TopicEvent event,
+                       TopicEventsListener.EventStage stage) {
+        notify(topic, event, stage, null);
+    }
+
+    /**
+     * Dispatches notification to all currently added listeners.
+     * @param topic
+     * @param event
+     * @param stage
+     * @param t
+     */
+    public void notify(String topic,
+                       TopicEventsListener.TopicEvent event,
+                       TopicEventsListener.EventStage stage,
+                       Throwable t) {
+        topicEventListeners
+                .forEach(listener -> notify(listener, topic, event, stage, t));
+    }
+
+    /**
+     * Dispatches SUCCESS/FAILURE notification to all currently added listeners on completion of the future.
+     * @param future
+     * @param topic
+     * @param event
+     * @param <T>
+     * @return future of a new completion stage
+     */
+    public <T> CompletableFuture<T> notifyOnCompletion(CompletableFuture<T> future,
+                                                       String topic,
+                                                       TopicEventsListener.TopicEvent event) {
+        return future.whenComplete((r, ex) -> notify(topic,
+                event,
+                ex == null ? TopicEventsListener.EventStage.SUCCESS : TopicEventsListener.EventStage.FAILURE,
+                ex));
+    }
+
+    /**
+     * Dispatches notification to specified listeners.
+     * @param listeners
+     * @param topic
+     * @param event
+     * @param stage
+     * @param t
+     */
+    public static void notify(TopicEventsListener[] listeners,
+                              String topic,
+                              TopicEventsListener.TopicEvent event,
+                              TopicEventsListener.EventStage stage,
+                              Throwable t) {
+        Objects.requireNonNull(listeners);
+        for (TopicEventsListener listener: listeners) {
+            notify(listener, topic, event, stage, t);
+        }
+    }
+
+    private static void notify(TopicEventsListener listener,
+                               String topic,
+                               TopicEventsListener.TopicEvent event,
+                               TopicEventsListener.EventStage stage,
+                               Throwable t) {
+        if (listener == null) {
+            return;
+        }
+
+        try {
+            listener.handleEvent(topic, event, stage, t);
+        } catch (Throwable ex) {
+            log.error("TopicEventsListener exception while handling {}_{} for topic {}", event, stage, topic, ex);
```

Comment:
We can print the `listener` so that if the `TopicEventListener` has implemented the `toString` method, we can know which `TopicEventListener` failed.